File: Python/NeetCode Blind 75/Two Pointers/PreReqs/TwoSum.py
# Given an array of integers nums and an integer target, 
# return indices of the two numbers such that they add up to target.

# You may assume that each input would have exactly one solution, and you may not use the same element twice.

# You can return the answer in any order.

 
# Example 1:
# Input: nums = [2,7,11,15], target = 9
# Output: [0,1]
# Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].

# Example 2:
# Input: nums = [3,2,4], target = 6
# Output: [1,2]

# Example 3:
# Input: nums = [3,3], target = 6
# Output: [0,1]

class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        res = []
        # L and R pointers
        # Start L at first index and R at len{nums} - 1
        # decrement R and compare to L if none match target continue
        # when we continue we increment L and then start R again from beginning
        L = 0
        R = len(nums) - 1
        while L < len(nums) - 1:
            if(R == L):
                L += 1
                R = len(nums) - 1
            if(nums[L] + nums[R] == target):
                res.append(L)
                res.append(R)
                return res
            if(R > L):
                R -= 1

        return res

    def run(self):
        # nums = [2,7,11,15]
        # target = 9
        # nums = [3,2,4]
        # target = 6
        nums = [3,3]
        target = 6
        print(self.twoSum(nums,target))


def main():
    solution = Solution()
    solution.run()

main()

File: C#/NeetCode Blind 150/ArraysAndHashing/twoSum.cs
// Given an array of integers nums and an integer target, 
// return the indices i and j such that nums[i] + nums[j] == target and i != j.

// You may assume that every input has exactly one pair of indices i and j that satisfy the condition.

// Return the answer with the smaller index first.

// Example 1:
// Input: 
// nums = [3,4,5,6], target = 7
// Output: [0,1]
// Explanation: nums[0] + nums[1] == 7, so we return [0, 1].

// Example 2:
// Input: nums = [4,5,6], target = 10
// Output: [0,2]

// Example 3:
// Input: nums = [5,5], target = 10
// Output: [0,1]


public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        // Use a Dict to store the values as we loop
        Dictionary<int,int> set = new Dictionary<int,int>();

        // Loop and subtract the target - num see if contains in Dictionary
        // If it does we add it to res
        for(int i=0; i<nums.Length; i++) {
            int diff = target - nums[i];
            // If target - num == dict.contains() add to res
            if(set.ContainsKey(target-nums[i])) {
                return new int[]{set[diff], i};
            }
            // Else we add to dictionary
            set[nums[i]] = i;
        }

        return new int[]{};
    }
}
